### DISCUSSION TITLE
Mui TextField value set by useState isn't registering in RHF for validation

### DISCUSSION THREAD

**Discussion by keithglassford** [react-hook-form/react-hook-form#8683] [Q&A]
**Describe the Issue**
I have a multistep form spread across three components which leverages the useFormContext and FormProvider to wrap the Materials-ui TextField inputs in each component of the Mui Stepper.

I am relatively new to react-hook-form (RHF) and Material-ui (Mui) and am trying to understand how to fix my issue which lies in the connections between the two. I know it's probably a simple fix and have gone through SO, the RHF closed issues but haven't been able to figure out viable solution. 

In the pared down example provided in the CSB link below, I show a mock of a single component  where another component on the page selects the text parameter and populates the TextField with the value with a useState. When the Submit is clicked, the form fields should be validated, in this case, that they are required.

To simplify the CSB example, I mocked the selection of the text field with useEffect which would run after page load to mimic a user's selection. While the TextField get's populated with "Test", when the submit button is pressed, it throws a validation error, indicating that the value hasn't yet been registered with RHF from the setter in the useState. 

If one clicks on the field and then clicks the Submit button, it validates correctly as the "Test" string get's registered with RHF. What I haven't been able to figure out yet after searching SO for hours, is what I am doing wrong.

**To Reproduce**

1. Once the CSB has loaded, you should see a "Test" string in the Text field.
2. Click Submit, which will show a validation error
3. Click in the text field and click Submit again, and the error will clear

**Codesandbox link (Required)**
https://codesandbox.io/s/brave-kowalevski-t53zj1?file=/src/App.js

**Expected behavior**
The form validation will recognize the updated value without having to activate the onChange to update the state.

**Additional context**
 I am using the most recent RHF v7.33.1

**Comment by keithglassford** [comment]
Ok, found a solution; although, there might be better way. I pulled the setValue from the methods of useForm, added the following to the useEffect to dynamically update the registered value given that the form has already been rendered:

setValue("text", "Test", { shouldValidate: true });

based upon the [docs](https://react-hook-form.com/api/useform/setvalue). In addition I moved register before onChange allow changes in the input field ... and it now works as expected.

**Comment by Moshyfawn** [comment]
There's `onChange` callback as a second argument options object for `register` to avoid overriding the `register` `onChange` one. Consider using it instead of an additional `useState`.

If you want to go for controlled input _(that has external value fed into your input)_, consider using `Controller` , `useController` API instead. Right now you're trying to workaround the `register` API behavior, that's designed to be used with uncontrolled inputs, aka the inputs where to source of truth for the input's value is the input itself.

Depending on your use-case, consider setting `mode: "onChange"` on `useForm` hook to validate on value change, like you trying to do with input example.

Here's [a very basic codesandbox](https://codesandbox.io/s/sharp-bell-s0i3d8?file=%2Fsrc%2FApp.js) to illustrate 2 approaches: controlled and uncontrolled input handling

**Reply by keithglassford** [reply]
Thanks for the suggestions!! I will give these a try.
